File: rust/ulsdk/src/request_context/api_key_context.rs
// Copyright (c), CommunityLogiq Software

use async_trait::async_trait;
use bytes::Bytes;
use ed25519_dalek::{Signature, Signer, SigningKey, SECRET_KEY_LENGTH};
use reqwest::header::{HeaderMap, HeaderName, HeaderValue};
use sha2::{Digest, Sha256};
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::time::sleep;
use url::form_urlencoded;
use uuid::Uuid;

use crate::error::Error;
use crate::keys::Key;
use crate::request_context::{
    get_endpoint, Environment, File, ParamMap, Region, RequestContext, DELAYS,
};

const REQUEST_TYPE: &str = "ul1_request";
const SIGNATURE_V1: &str = "UL1-ED25519";
const HEADER_AUTHORIZATION: &str = "authorization";
const HEADER_X_UL_DATE: &str = "x-ul-date";
const HEADER_CONTENT_TYPE: &str = "content-type";
const SIGNED_HEADERS: &[&str] = &[HEADER_X_UL_DATE];

#[derive(Clone)]
pub struct ApiKeyContext {
    environment: Environment,
    key: Key,
}

impl ApiKeyContext {
    pub fn new(key: Key, environment: Environment) -> Self {
        Self { environment, key }
    }

    pub fn user_id(&self) -> Uuid {
        self.key.user_id()
    }
}

fn canonicalize_path(path: &str) -> String {
    if path.is_empty() {
        return "/".to_owned();
    }

    if !path.starts_with('/') {
        return format!("/{}", path);
    }

    path.to_owned()
}

fn canonicalize_query_string(query: &mut [(String, String)]) -> String {
    if query.is_empty() {
        return String::new();
    }

    query.sort_by(|a, b| a.0.cmp(&b.0));

    let components = query
        .iter()
        .filter(|(k, _)| k != "X-UL-Signature")
        .map(|(k, v)| {
            format!(
                "{}={}",
                k,
                form_urlencoded::byte_serialize(v.as_bytes()).collect::<String>()
            )
        })
        .collect::<Vec<_>>();
    components.join("&")
}

fn canonicalize_headers(signed_headers: &[&str], headers: &HeaderMap) -> Result<String, Error> {
    let mut sorted_signed_headers = signed_headers
        .iter()
        .map(|x| x.to_lowercase())
        .collect::<Vec<_>>();
    sorted_signed_headers.sort();

    let canonical_header_parts = sorted_signed_headers
        .into_iter()
        .flat_map(|v| {
            match HeaderName::from_lowercase(v.as_bytes()).map_err(|e| Error::from(e.to_string())) {
                Ok(header_name) => {
                    if let Some(value) = headers.get(header_name) {
                        match value.to_str().map_err(|e| Error::from(e.to_string())) {
                            Ok(value) => Some(Ok(format!("{}:{}", v, value.trim()))),
                            Err(e) => Some(Err(e)),
                        }
                    } else {
                        None
                    }
                }
                Err(e) => Some(Err(e)),
            }
        })
        .collect::<Result<Vec<_>, _>>()?;

    Ok(canonical_header_parts.join("\n"))
}

fn hash(slice: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(slice);
    hex::encode(hasher.finalize())
}

fn canonicalize_request(
    method: &str,
    path: &str,
    query: &mut [(String, String)],
    headers: &HeaderMap,
    signed_headers: &[&str],
    body: &[u8],
) -> Result<(String, String), Error> {
    let canonical_path = canonicalize_path(path);
    let canonical_query_string = canonicalize_query_string(query);
    let canonical_headers = canonicalize_headers(signed_headers, headers)?;
    let signed_headers = signed_headers.to_vec().join(";");

    let s = format!(
        "{}\n{}\n{}\n{}\n{}\n{}",
        method.to_uppercase(),
        canonical_path,
        canonical_query_string,
        canonical_headers,
        signed_headers,
        hash(body)
    );

    Ok((hash(s.as_bytes()), signed_headers))
}

fn generate_auth_header(
    key: &Key,
    method: &str,
    path: &str,
    query: &mut [(String, String)],
    headers: &mut HeaderMap,
    body: &[u8],
) -> Result<String, Error> {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs()
        .to_string();

    headers.insert(
        HeaderName::from_static(HEADER_X_UL_DATE),
        HeaderValue::from_str(&ts).unwrap(),
    );

    let (request_hash, signed_headers) =
        canonicalize_request(method, path, query, headers, SIGNED_HEADERS, body)?;

    let scope = format!("{}/{}/{}/{}", key.user_id(), ts, key.region(), REQUEST_TYPE);
    let signing_string = format!("{}\n{}\n{}", SIGNATURE_V1, scope, request_hash);

    let secret_key = key.secret_key()[0..SECRET_KEY_LENGTH].try_into().unwrap();
    let keypair = SigningKey::from_bytes(secret_key);
    let signature: Signature = keypair.sign(signing_string.as_bytes());
    let signature = hex::encode(signature.to_bytes());

    Ok(format!(
        "{} Credential={}/{}, SignedHeaders={}, Signature={}",
        SIGNATURE_V1,
        key.access_key(),
        scope,
        signed_headers,
        signature
    ))
}

#[async_trait]
impl RequestContext for ApiKeyContext {
    fn region(&self) -> Region {
        self.key.region()
    }

    fn environment(&self) -> Environment {
        self.environment
    }

    async fn get(
        &self,
        path: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let endpoint = get_endpoint(self.key.region(), self.environment, path);
        let client = reqwest::Client::new();
        let mut response = None;

        let params = params.unwrap_or_default();

        let mut query = Vec::new();
        params
            .iter()
            .for_each(|(k, v)| query.push((k.to_owned(), v.to_owned())));

        let mut headers = headers.unwrap_or_default();
        let auth_header =
            generate_auth_header(&self.key, "GET", path, &mut query, &mut headers, &[])?;

        headers.insert(
            HeaderName::from_static(HEADER_AUTHORIZATION),
            HeaderValue::from_str(&auth_header).unwrap(),
        );

        for delay in DELAYS {
            let r = client
                .get(&endpoint)
                .headers(headers.clone())
                .query(&params)
                .send()
                .await?;

            if r.status().is_server_error() {
                response = Some(r);
                sleep(Duration::from_millis(*delay)).await;
            } else {
                response = Some(r);
                break;
            }
        }

        let response = if let Some(r) = response {
            r
        } else {
            return Err(Error::Unclassified("No error from endpoint".into()));
        };

        if !response.status().is_success() {
            let status = response.status();
            let bytes = response.bytes().await?;
            return Err(Error::FailedRequest(
                endpoint.to_owned(),
                "Unable to retrieve data from endpoint".to_owned(),
                status,
                unsafe { std::str::from_utf8_unchecked(&bytes) }.to_owned(),
            ));
        }

        response
            .bytes()
            .await
            .map_err(Error::ReqwestError)
            .map(|b| b.to_vec())
    }

    async fn put(
        &self,
        path: &str,
        body: Bytes,
        mimetype: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let endpoint = get_endpoint(self.key.region(), self.environment, path);
        let client = reqwest::Client::new();
        let mut response = None;

        let params = params.unwrap_or_default();

        let mut query = Vec::new();
        params
            .iter()
            .for_each(|(k, v)| query.push((k.to_owned(), v.to_owned())));

        let mut headers = headers.unwrap_or_default();
        headers.insert(
            HeaderName::from_static(HEADER_CONTENT_TYPE),
            HeaderValue::from_str(mimetype).unwrap(),
        );

        let auth_header =
            generate_auth_header(&self.key, "PUT", path, &mut query, &mut headers, &body)?;

        headers.insert(
            HeaderName::from_static(HEADER_AUTHORIZATION),
            HeaderValue::from_str(&auth_header).unwrap(),
        );

        for delay in DELAYS {
            let r = client
                .put(&endpoint)
                .headers(headers.clone())
                .body(body.clone())
                .send()
                .await?;

            if r.status().is_server_error() {
                response = Some(r);
                sleep(Duration::from_millis(*delay)).await;
            } else {
                response = Some(r);
                break;
            }
        }

        let response = if let Some(r) = response {
            r
        } else {
            return Err(Error::Unclassified("No error from endpoint".into()));
        };

        if !response.status().is_success() {
            let status = response.status();
            let bytes = response.bytes().await?;
            return Err(Error::FailedRequest(
                endpoint.to_owned(),
                "Unable to retrieve data from endpoint".to_owned(),
                status,
                unsafe { std::str::from_utf8_unchecked(&bytes) }.to_owned(),
            ));
        }

        response
            .bytes()
            .await
            .map_err(Error::ReqwestError)
            .map(|b| b.to_vec())
    }

    async fn post(
        &self,
        path: &str,
        body: Bytes,
        mimetype: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let endpoint = get_endpoint(self.key.region(), self.environment, path);
        let client = reqwest::Client::new();
        let mut response = None;

        let params = params.unwrap_or_default();

        let mut query = Vec::new();
        params
            .iter()
            .for_each(|(k, v)| query.push((k.to_owned(), v.to_owned())));

        let mut headers = headers.unwrap_or_default();
        headers.insert(
            HeaderName::from_static(HEADER_CONTENT_TYPE),
            HeaderValue::from_str(mimetype).unwrap(),
        );

        let auth_header =
            generate_auth_header(&self.key, "POST", path, &mut query, &mut headers, &body)?;

        headers.insert(
            HeaderName::from_static(HEADER_AUTHORIZATION),
            HeaderValue::from_str(&auth_header).unwrap(),
        );

        for delay in DELAYS {
            let r = client
                .post(&endpoint)
                .headers(headers.clone())
                .body(body.clone())
                .send()
                .await?;

            if r.status().is_server_error() {
                response = Some(r);
                sleep(Duration::from_millis(*delay)).await;
            } else {
                response = Some(r);
                break;
            }
        }

        let response = if let Some(r) = response {
            r
        } else {
            return Err(Error::Unclassified("No error from endpoint".into()));
        };

        if !response.status().is_success() {
            let status = response.status();
            let bytes = response.bytes().await?;
            return Err(Error::FailedRequest(
                endpoint.to_owned(),
                "Unable to retrieve data from endpoint".to_owned(),
                status,
                unsafe { std::str::from_utf8_unchecked(&bytes) }.to_owned(),
            ));
        }

        response
            .bytes()
            .await
            .map_err(Error::ReqwestError)
            .map(|b| b.to_vec())
    }

    async fn upload(&self, _path: &str, _files: Vec<File>) -> Result<(), Error> {
        todo!()
    }

    async fn delete(
        &self,
        path: &str,
        params: Option<ParamMap>,
        headers: Option<HeaderMap>,
    ) -> Result<Vec<u8>, Error> {
        let endpoint = get_endpoint(self.key.region(), self.environment, path);
        let client = reqwest::Client::new();
        let mut response = None;

        let params = params.unwrap_or_default();

        let mut query = Vec::new();
        params
            .iter()
            .for_each(|(k, v)| query.push((k.to_owned(), v.to_owned())));

        let mut headers = headers.unwrap_or_default();
        let auth_header =
            generate_auth_header(&self.key, "GET", path, &mut query, &mut headers, &[])?;

        headers.insert(
            HeaderName::from_static(HEADER_AUTHORIZATION),
            HeaderValue::from_str(&auth_header).unwrap(),
        );

        for delay in DELAYS {
            let r = client
                .delete(&endpoint)
                .headers(headers.clone())
                .query(&params)
                .send()
                .await?;

            if r.status().is_server_error() {
                response = Some(r);
                sleep(Duration::from_millis(*delay)).await;
            } else {
                response = Some(r);
                break;
            }
        }

        let response = if let Some(r) = response {
            r
        } else {
            return Err(Error::Unclassified("No error from endpoint".into()));
        };

        if !response.status().is_success() {
            let status = response.status();
            let bytes = response.bytes().await?;
            return Err(Error::FailedRequest(
                endpoint.to_owned(),
                "Unable to retrieve data from endpoint".to_owned(),
                status,
                unsafe { std::str::from_utf8_unchecked(&bytes) }.to_owned(),
            ));
        }

        response
            .bytes()
            .await
            .map_err(Error::ReqwestError)
            .map(|b| b.to_vec())
    }
}

File: go/api_key_context.go
// Copyright (c), CommunityLogiq Software

package ulsdk

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"
)

const (
	requestType = "ul1_request"
	signatureV1 = "UL1-ED25519"
)

// ApiKeyContext implements the api.RequestContext interface using ED25519-signed
// API key authentication.
type ApiKeyContext struct {
	key         *Key
	environment Environment
	client      *http.Client
}

// NewApiKeyContext creates a new ApiKeyContext with the given key and environment.
func NewApiKeyContext(key *Key, environment Environment) *ApiKeyContext {
	return &ApiKeyContext{
		key:         key,
		environment: environment,
		client:      &http.Client{},
	}
}

func canonicalizePath(path string) string {
	if strings.HasPrefix(path, "/") {
		return path
	}
	return "/" + path
}

func canonicalizeQueryString(params [][2]string) string {
	if len(params) == 0 {
		return ""
	}

	var components []string
	for _, p := range params {
		if p[0] == "X-UL-Signature" {
			continue
		}
		components = append(components, p[0]+"="+url.QueryEscape(p[1]))
	}
	sort.Strings(components)
	return strings.Join(components, "&")
}

func canonicalizeHeaders(signedHeaders []string, headers map[string]string) string {
	sorted := make([]string, len(signedHeaders))
	copy(sorted, signedHeaders)
	sort.Strings(sorted)

	var parts []string
	for _, h := range sorted {
		parts = append(parts, strings.ToLower(h)+":"+strings.TrimSpace(headers[h]))
	}
	return strings.Join(parts, "\n")
}

func hashBytes(data []byte) string {
	h := sha256.Sum256(data)
	return fmt.Sprintf("%x", h[:])
}

func canonicalizeRequest(
	method string,
	path string,
	params [][2]string,
	headers map[string]string,
	signedHeaders []string,
	body []byte,
) string {
	canonicalPath := canonicalizePath(path)
	canonicalQuery := canonicalizeQueryString(params)
	canonicalHdrs := canonicalizeHeaders(signedHeaders, headers)

	sorted := make([]string, len(signedHeaders))
	copy(sorted, signedHeaders)
	sort.Strings(sorted)

	s := method + "\n" +
		canonicalPath + "\n" +
		canonicalQuery + "\n" +
		canonicalHdrs + "\n" +
		strings.Join(sorted, ";") + "\n" +
		hashBytes(body)

	return hashBytes([]byte(s))
}

func generateAuthHeader(
	key *Key,
	method string,
	path string,
	params [][2]string,
	headers map[string]string,
	body []byte,
) {
	ts := time.Now().Unix()
	signedHeaders := []string{"x-ul-date"}
	headers["x-ul-date"] = fmt.Sprintf("%d", ts)

	requestHash := canonicalizeRequest(method, path, params, headers, signedHeaders, body)
	scope := fmt.Sprintf("%s/%d/%s/%s", key.UserID, ts, key.Region, requestType)

	signingString := signatureV1 + "\n" + scope + "\n" + requestHash

	// Decode the base64-encoded secret key (with padding)
	secretKeyBytes, err := base64.StdEncoding.DecodeString(key.SecretKey + "==")
	if err != nil {
		// Try without extra padding
		secretKeyBytes, _ = base64.StdEncoding.DecodeString(key.SecretKey)
	}

	// Use the first 32 bytes as the ED25519 seed
	seed := secretKeyBytes[:32]
	privateKey := ed25519.NewKeyFromSeed(seed)
	signature := ed25519.Sign(privateKey, []byte(signingString))

	authHeader := fmt.Sprintf(
		"%s Credential=%s/%s, SignedHeaders=%s, Signature=%x",
		signatureV1,
		key.AccessKey,
		scope,
		strings.Join(signedHeaders, ";"),
		signature,
	)
	headers["authorization"] = authHeader
}

func (c *ApiKeyContext) doRequest(method, path string, body []byte, contentType string, params [][2]string, headers map[string]string) ([]byte, error) {
	// Merge caller-provided headers into a mutable copy
	mergedHeaders := make(map[string]string)
	for k, v := range headers {
		mergedHeaders[k] = v
	}

	if contentType != "" {
		mergedHeaders["content-type"] = contentType
	}

	generateAuthHeader(c.key, method, path, params, mergedHeaders, body)

	endpoint := GetEndpoint(c.key.Region, c.environment, path)

	// Append query parameters
	if len(params) > 0 {
		q := url.Values{}
		for _, p := range params {
			q.Add(p[0], p[1])
		}
		endpoint = endpoint + "?" + q.Encode()
	}

	var bodyReader io.Reader
	if body != nil {
		bodyReader = strings.NewReader(string(body))
	}

	req, err := http.NewRequest(method, endpoint, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	for k, v := range mergedHeaders {
		req.Header.Set(k, v)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(respBody))
	}

	return respBody, nil
}

// Get makes an authenticated GET request.
func (c *ApiKeyContext) Get(path string, params [][2]string, headers map[string]string) ([]byte, error) {
	return c.doRequest("GET", path, nil, "", params, headers)
}

// Post makes an authenticated POST request.
func (c *ApiKeyContext) Post(path string, body []byte, contentType string, params [][2]string, headers map[string]string) ([]byte, error) {
	return c.doRequest("POST", path, body, contentType, params, headers)
}

// Put makes an authenticated PUT request.
func (c *ApiKeyContext) Put(path string, body []byte, contentType string, params [][2]string, headers map[string]string) ([]byte, error) {
	return c.doRequest("PUT", path, body, contentType, params, headers)
}

// Delete makes an authenticated DELETE request.
func (c *ApiKeyContext) Delete(path string, params [][2]string, headers map[string]string) ([]byte, error) {
	return c.doRequest("DELETE", path, nil, "", params, headers)
}
